cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+
+	// We need to check whether pods marked as unschedulable are actually unschedulable.
+	// It's likely we added a new node and the scheduler just haven't managed to put the
+	// pod on in yet. In this situation we don't want to trigger another scale-up.
+	//
+	// It's also important to prevent uncontrollable cluster growth if CA's simulated
+	// scheduler differs in opinion with real scheduler. Example of such situation:
+	// - CA and Scheduler has slightly different configuration
+	// - Scheduler can't schedule a pod and marks it as unschedulable
+	// - CA added a node which should help the pod
+	// - Scheduler doesn't schedule the pod on the new node
+	//   because according to it logic it doesn't fit there
+	// - CA see the pod is still unschedulable, so it adds another node to help it
+	//
+	// With the check enabled the last point won't happen because CA will ignore a pod
+	// which is supposed to schedule on an existing node.
+	ls.scaleDownForbidden = false
+
+	unschedulablePodsWithoutTPUs := tpu.ClearTPURequests(allUnschedulablePods)
+
+	// Some unschedulable pods can be waiting for lower priority pods preemption so they have nominated node to run.
+	// Such pods don't require scale up but should be considered during scale down.
+	var unschedulablePods []*apiv1.Pod
+	unschedulablePods, ls.unschedulableWaitingForLowerPriorityPreemption = FilterOutExpendableAndSplit(
+		unschedulablePodsWithoutTPUs, ls.autoscaler.ExpendablePodsPriorityCutoff)
+
+	glog.V(4).Infof("Filtering out schedulables")
+	filterOutSchedulableStart := time.Now()
+	ls.unschedulablePodsToHelp = FilterOutSchedulable(unschedulablePods, ls.readyNodes, ls.allScheduled,
+		ls.unschedulableWaitingForLowerPriorityPreemption, ls.autoscaler.PredicateChecker, ls.autoscaler.ExpendablePodsPriorityCutoff)
+	metrics.UpdateDurationFromStart(metrics.FilterOutSchedulable, filterOutSchedulableStart)
+
+	if len(ls.unschedulablePodsToHelp) != len(unschedulablePods) {
+		glog.V(2).Info("Schedulable pods present")
+		ls.scaleDownForbidden = true
+	} else {
+		glog.V(4).Info("No schedulable pods")
+	}
+
+	return ls.onSuccess()
+}
+
+// See LoopStateHandler interface
+func (ls *loopState) ScaleUp() (ExecutionResult, errors.AutoscalerError) {
+	if len(ls.unschedulablePodsToHelp) == 0 {
+		glog.V(1).Info("No unschedulable pods")
+	} else if ls.autoscaler.MaxNodesTotal > 0 && len(ls.readyNodes) >= ls.autoscaler.MaxNodesTotal {
+		glog.V(1).Info("Max total nodes in cluster reached")
+	} else if allPodsAreNew(ls.unschedulablePodsToHelp, ls.currentTime) {
+		// The assumption here is that these pods have been created very recently and probably there
+		// is more pods to come. In theory we could check the newest pod time but then if pod were created
+		// slowly but at the pace of 1 every 2 seconds then no scale up would be triggered for long time.
+		// We also want to skip a real scale down (just like if the pods were handled).
+		ls.scaleDownForbidden = true
```

Comment:
If the purpose of this change is splitting the code by logic, I would expect that:
- ScaleUp() method only deals with scale-up, 
- ScaleDown() only deals with scale-down,
- whatever method calls ScaleUp and/or ScaleDown should decide when/if to call them.

Right now, both ScaleUp() and PreparePodData() explicitly decide that scale-down is forbidden. The logic is still all over the place, but now the names suggest it isn't (making it harder to find out what happens.)

Reply:
A very valid point. As is evident by the sheer size of ScaleUp and ScaleDown methods they were refactored the least and, as you observed, contain unwanted interactions from the code prior to refactoring.

I think I can improve the situation a bit here, avoiding `scaleDownForbidden` flag but I'm afraid some entanglement issues will still be left.
TODO(kgolab) - revisit this